File: code/max_product3/max_product3.cpp
#include <iostream>
#include <vector>

#include <chrono>
#include <ctime>    

using namespace std;

int main(void)
{
    auto start = std::chrono::system_clock::now();

    int size;
    int value;

    cin >> size;

    if (size < 3)
    {
        throw std::invalid_argument("Data array must contain at least 3 values");
    }

    vector<int> data(size);

    int i = 0;
    while (i < size)
    {
        cin >> value;
        data[i] = value;
        i++;
    }      
    
    auto middle = std::chrono::system_clock::now();

    int max1 = data[0];
    int max2 = data[1];
    int max3 = data[2];
    
    int tmp;
    
    if (max1 < max2)
    {
        tmp = max1;
        max1 = max2;
        max2 = tmp;
    }

    if (max1 < max3)
    {
        tmp = max1;
        max1 = max3;
        max3 = tmp;
    }
            
    if (max2 < max3)
    {
        tmp = max2;
        max2 = max3;
        max3 = tmp;
    }
                
    int min1 = max3;
    int min2 = max2;

    int idx = 3;
    while (idx < data.size())
    {
        if (data[idx] > max1)
        {
            max3 = max2;
            max2 = max1;
            max1 = data[idx];
        }
        else if (data[idx] > max2)
        {
            max3 = max2;
            max2 = data[idx];
        }
        else if (data[idx] > max3)
        {
            max3 = data[idx];
        };
        
        if (data[idx] < min1)
        {
            min2 = min1;
            min1 = data[idx];
        }
        else if (data[idx] < min2)
        {
            min2 = data[idx];
        };

        idx++;
    }

    long long m1 = max1;
    m1 *= max2;
    m1 *= max3;

    long long m2 = min1;
    m2 *= min2;
    m2 *= max1;

    if (m1 > m2)
    { 
        cout << m1;
    }
    else
    {
        cout << m2;
    }

    auto end = std::chrono::system_clock::now();

    std::chrono::duration<double> elapsed_seconds = end - start;
    std::chrono::duration<double> elapsed_seconds_mid = middle - start;
    
    std::time_t end_time = std::chrono::system_clock::to_time_t(end);

    std::cout << endl << "elapsed time: " << elapsed_seconds.count() << "s" << std::endl;
    std::cout << "elapsed time load: " << elapsed_seconds_mid.count() << "s" << std::endl;
    
    return 0;
}

File: code/max_product3/max_product3.py
import time

def readData(fileName = './code/max_product3/test_0.txt'):
    data = list()
    
    f = open(fileName,'r')
    try:
        data_size = int(f.readline())
        while True:
            if (len(data) == data_size):
                break   
            res = int(f.readline())   
            data.append(res)
    finally:
        f.close()
    return data
    
def getMaxProduct3_sort(data: list):
    data.sort() # N * LOG(N)

    max1 = data[len(data)-1] 
    max2 = data[len(data)-2] 
    max3 = data[len(data)-3] 
    min1 = data[0]
    min2 = data[1]

    return max(max1 * max2 * max3, min1 * min2 * max1)

def getMaxProduct3_line(data: list):
    if len(data) < 3:
        raise ValueError('Data array must contain at least 3 values')

    max1 = data[0] 
    max2 = data[1] 
    max3 = data[2] 

    if max1 < max2:
        max1, max2 = max2, max1
    if max1 < max3:
        max1, max3 = max3, max1
    if max2 < max3:
        max2, max3 = max3, max2

    min1 = max3
    min2 = max2

    # invariant: i=3; max1, max2, max3 - max of three, min1, min2 - min of three

    idx = 3
    while idx < len(data):
        if data[idx] > max1:
            max1, max2, max3 = data[idx], max1, max2
        elif data[idx] > max2:
            max2, max3 = data[idx], max2
        elif data[idx] > max3:
            max3 = data[idx]

        if data[idx] < min1:
            min1, min2 = data[idx], min1
        elif data[idx] < min2:
            min2 = data[idx]

        idx += 1

    return max(max1 * max2 * max3, min1 * min2 * max1)

if __name__ == "__main__":
    dir = './code/max_product3/'
    
    t1 = time.time()
    
    data = readData(dir + 'test_5.txt')   
    
    # res = getMaxProduct3_sort(data)
    res = getMaxProduct3_line(data)

    t2 = time.time()

    # print('time=', t2-t1)
    print(res)
    
